File: secondbest-tauri/src/types/game.ts
// Second Best Game - フロントエンド型定義

export enum Position {
  N = "N",
  NE = "NE", 
  E = "E",
  SE = "SE",
  S = "S",
  SW = "SW",
  W = "W", 
  NW = "NW"
}

export enum Player {
  Black = "Black",
  White = "White"
}

export interface PieceStack {
  pieces: Player[]; // 最大3個まで、先頭が一番下
}

export enum TurnPhase {
  WaitingForMove = "WaitingForMove",           // 通常の手番待ち
  WaitingForSecondBest = "WaitingForSecondBest", // Second Best宣言待ち
  WaitingForSecondMove = "WaitingForSecondMove"   // Second Best後の代替手待ち
}

export interface GameState {
  board: Record<Position, PieceStack>;
  current_player: Player;
  turn_phase: TurnPhase;
  second_best_available: boolean;
  winner?: Player;
}

export type MoveAction = 
  | { Place: { position: Position; player: Player } }
  | { Move: { from: Position; to: Position } };

export enum GameOverReason {
  VerticalLineup = "VerticalLineup",
  HorizontalLineup = "HorizontalLineup",
  NoMoves = "NoMoves"
}

// Push型APIのイベント構造体
export interface AiMoveEvent {
  action: MoveAction;
  new_state: GameState;
}

export interface AiSecondBestEvent {
  new_state: GameState;
}

export interface AiSecondMoveEvent {
  action: MoveAction;
  new_state: GameState;
}

export interface GameOverEvent {
  winner?: Player;
  reason: GameOverReason;
}

export interface TurnPhaseEvent {
  new_phase: TurnPhase;
  current_player: Player;
}

export interface AiErrorEvent {
  message: string;
}

// Pull型API関数の型定義
export interface GameAPI {
  // ゲーム管理
  new_game(): Promise<GameState>;
  get_game_state(): Promise<GameState>;
  get_legal_moves(): Promise<MoveAction[]>;
  
  // プレイヤーアクション
  make_move(action: MoveAction): Promise<GameState>;
  declare_second_best(): Promise<GameState>;
  
  // ゲーム情報
  check_winner(): Promise<Player | null>;
  get_position_stack(position: Position): Promise<PieceStack>;
  can_declare_second_best(): Promise<boolean>;
} 

File: secondbest-tauri/src-tauri/src/game.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum Position {
    N,
    NE,
    E,
    SE,
    S,
    SW,
    W,
    NW,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum Player {
    Black,
    White,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PieceStack {
    pub pieces: Vec<Player>, // 最大3個まで、先頭が一番下
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum TurnPhase {
    WaitingForMove,       // 通常の手番待ち
    WaitingForSecondBest, // Second Best宣言待ち
    WaitingForSecondMove, // Second Best後の代替手待ち
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GameState {
    pub board: HashMap<Position, PieceStack>,
    pub current_player: Player,
    pub turn_phase: TurnPhase,
    pub second_best_available: bool,
    pub winner: Option<Player>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum MoveAction {
    Place { position: Position, player: Player },
    Move { from: Position, to: Position },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum GameOverReason {
    VerticalLineup,
    HorizontalLineup,
    NoMoves,
}

// Push型APIのイベント構造体
#[derive(Debug, Clone, Serialize)]
pub struct AiMoveEvent {
    pub action: MoveAction,
    pub new_state: GameState,
}

#[derive(Debug, Clone, Serialize)]
pub struct AiSecondBestEvent {
    pub new_state: GameState,
}

#[derive(Debug, Clone, Serialize)]
pub struct AiSecondMoveEvent {
    pub action: MoveAction,
    pub new_state: GameState,
}

#[derive(Debug, Clone, Serialize)]
pub struct GameOverEvent {
    pub winner: Option<Player>,
    pub reason: GameOverReason,
}

#[derive(Debug, Clone, Serialize)]
pub struct TurnPhaseEvent {
    pub new_phase: TurnPhase,
    pub current_player: Player,
}

#[derive(Debug, Clone, Serialize)]
pub struct AiErrorEvent {
    pub message: String,
}
